### PR TITLE
38 stratified dataset split

### PR DESCRIPTION
**Pull request by Phaneron97**
# What does it require?
`b_o_splitter.py` (see #39) **_has to be run at least once_** to move all files from their original into 2 folders called "b" (above rooftiles) and "o" (below rooftiles)

# What does it do?
* Makes a copy of "b" and "o" (doesn't do anything to contents of "b" and "o" to preserve the original dataset if required) 
* Uses `sklearn`'s `train_test_split` to stratify split the data into "test" and "train" folders
* When run again when for example you want a different `train_ratio` (which can be set in `if __name__ == "__main__":`) , the existing files will get shuffled again into "train" and "test" folders. 
* No new files will be made when code is run again

### CODE REVIEW

File: code/data_analysis/stratified_splitter.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                    os.makedirs(os.path.dirname(destination_path), exist_ok=True)
+                    shutil.copy(source_path, destination_path)
+                    print(f"Copied '{file}' from '{class_folder}' to 'train/{class_folder}/{os.path.basename(subdir)}'")
+
+                for file in test_files:
+                    source_path = os.path.join(subdir, file)
+                    destination_path = os.path.join(test_folder, os.path.basename(subdir), file)
+                    os.makedirs(os.path.dirname(destination_path), exist_ok=True)
+                    shutil.copy(source_path, destination_path)
+                    print(f"Copied '{file}' from '{class_folder}' to 'test/{class_folder}/{os.path.basename(subdir)}'")
```
Comment (Bastiaan27) [inline code/data_analysis/stratified_splitter.py:44]: I see repetitive code here, is there a way to simplify this?

File: code/data_analysis/stratified_splitter.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        test_folder = os.path.join(source_folder, "test", class_folder)
+
+        # Clear existing train and test folders
+        if os.path.exists(train_folder):
+            shutil.rmtree(train_folder)
+        os.makedirs(train_folder)
+
+        if os.path.exists(test_folder):
+            shutil.rmtree(test_folder)
+        os.makedirs(test_folder)
```
Comment (Bastiaan27) [inline code/data_analysis/stratified_splitter.py:26]: This could be a separate method, you should be immediately able to know what a method does by reading the title

**Review by Bastiaan27 — CHANGES_REQUESTED** [review]
What I see in this PR is really good, just some small changes left to make

**Comment by Phaneron97** [comment]
> Does it put the copies in a different folder? Isn't it better to backup the whole dataset before you split the top and bottom pictures of the rooftiles just like we do with the other data cleaning methods? Otherwise we'll just end up with a lot of extra data we don't need

@Bastiaan27 The test and train data will be put in the same folder where the original "b" and "o" files kept. The end result will be like this:

- Dataset
    - o
        - d0000_o_blw_g_OVH
        - d0001_o_blw_n_TDN44
        - d0002_o_nrd_g_VH
    - b
        - d0000_b_blw_g_OVH
        - d0001_b_blw_n_TDN44
        - d0002_b_nrd_g_VH
    - train
        - b
            - d0000_b_blw_g_OVH
            - d0001_b_blw_n_TDN44
            - d0002_b_nrd_g_VH
        - o 
            - d0000_o_blw_g_OVH
            - d0001_o_blw_n_TDN44
            - d0002_o_nrd_g_VH
    - test
         - b
            - d0000_b_blw_g_OVH
            - d0001_b_blw_n_TDN44
            - d0002_b_nrd_g_VH
        - o 
            - d0000_o_blw_g_OVH
            - d0001_o_blw_n_TDN44
            - d0002_o_nrd_g_VH

File: code/data_analysis/stratified_splitter.py (hunk 2 above)
Comment (Phaneron97) [reply code/data_analysis/stratified_splitter.py:26]: I agree, I'll change this.

File: code/data_analysis/stratified_splitter.py (hunk 1 above)
Comment (Phaneron97) [reply code/data_analysis/stratified_splitter.py:44]: I added a new method `copy_to_train_test_folder` which takes the train and test files, their folders, the original folder and file, and the subdirectory which will be something like `d0002_o_nrd_g_VH` (for example)

**Comment by Phaneron97** [comment]
@Bastiaan27 I updated the code, is it okay now?

**Review by Bastiaan27 — APPROVED** [review]
Looks good! Code is nice and compact now, well done!

**Comment by Bastiaan27** [comment]
jojo-go-ahead

**Comment by Bastiaan27** [comment]
> > Does it put the copies in a different folder? Isn't it better to backup the whole dataset before you split the top and bottom pictures of the rooftiles just like we do with the other data cleaning methods? Otherwise we'll just end up with a lot of extra data we don't need
> 
> @Bastiaan27 The test and train data will be put in the same folder where the original "b" and "o" files kept. The end result will be like this:
> 
> * Dataset
>   
>   * o
>     
>     * d0000_o_blw_g_OVH
>     * d0001_o_blw_n_TDN44
>     * d0002_o_nrd_g_VH
>   * b
>     
>     * d0000_b_blw_g_OVH
>     * d0001_b_blw_n_TDN44
>     * d0002_b_nrd_g_VH
>   * train
>     
>     * b
>       
>       * d0000_b_blw_g_OVH
>       * d0001_b_blw_n_TDN44
>       * d0002_b_nrd_g_VH
>     * o
>       
>       * d0000_o_blw_g_OVH
>       * d0001_o_blw_n_TDN44
>       * d0002_o_nrd_g_VH
>   * test
>     
>     * b
>       
>       * d0000_b_blw_g_OVH
>       * d0001_b_blw_n_TDN44
>       * d0002_b_nrd_g_VH
>     * o
>       
>       * d0000_o_blw_g_OVH
>       * d0001_o_blw_n_TDN44
>       * d0002_o_nrd_g_VH

Ah okay, that looks clear
